Source/WebCore/loader/ImageLoader.cpp

```diff
@@ -486,18 +487,32 @@ void ImageLoader::notifyFinished(CachedResource& resource, const NetworkLoadMetr
         return;
     }
 
-    if (hasPendingDecodePromises())
-        decode();
-    loadEventSender().dispatchEventSoon(*this, eventNames().loadEvent);
+    Ref doc = document();
+    setShouldDelayLoadEvent(doc, true);
+    m_image->protectedImage()->subresourcesAreFinished([this, protectedThis = Ref { *this }, doc]() mutable {
```

Comment:
Maybe `subresourcesAreFinished` should be the one doing the incrementing at call site/decrementing just before calling the completion handler.
This way, no need to add a new `setShouldDelayLoadEvent` virtual method.

Maybe `subresourcesAreFinished` could take a ShouldDelayLoadEvent boolean enum.